README.md
# Single page app: vanilla JS(frontend) + JSON REST API with mongo & express(backend)

### Created by

View README.md of https://github.com/darco2018/mustwatch-REST-API-mongo-express to see how the API part was created.

```sh
$ git clone  https://github.com/darco2018/mustwatch-REST-API-mongo-express.git
$ npm i
$ mkdir views public/stylesheets public/javascripts public/images
```

- create **/views/index.html**
- ###### go to the INSTALLATION section down below & choose one-server or two-server option

* **MONGO**

  - install mongo & mongo CLI on your machine
  - start mongo demon & mongo:

```sh
$ sudo service mongod start/status/stop
$ mongo
```

- uncomment dbTest.test(mongoose) in app.js & check it's working

### Time saving tips

- when fetching all the movies, note that what is returned from res.json() is not an array but **{**movies: Array(1)**}**. To get the movies you have to **res.json().movies**
- nodemon restarts the server on changes but **doesn't** refresh the page in browser (use LIVE SERVER for hot reload)

### Installation

###### You can run frontend & backend on one server or on two servers. The benefit of using the latter option is **`HOT RELOAD`**.

&NewLine;

##### Single server, single port.

###### Just Nodemon with the fronend & the backend at 3000.

&NewLine;

```sh
$ npm i
$ npm run debug
// go to  http://localhost:3000/ & http://localhost:3000/api/movies
$ sudo kill -9 $(sudo lsof -t -i:3000) // will stop the server
```

1. For the css & js links in your **index.html** use **relative paths to /public** folder, eg
   <link rel="stylesheet" href="stylesheets/styles.css" >
   <script src="javascripts/index.js"></script>
2. In **app.js**:
   app.use(express.static(path.join(**dirname + '/views')));
   app.use(express.static(path.join(**dirname, '/public')));
3. in the **index route** change
   res.send to res.sendFile('index.html');

##### Two servers, two ports:

###### Nodemon will take care of the backend at 3000. Live Server will deal with frontend at 5500.

&NewLine;

1. Backend the same as above but add the following line to **package.json**:
   "`norestart`": "nodemon ./bin/www --ignore '/\*' " - it will prevent useless backend server reloads (You want to work on the fronend, don't you?!)
   &NewLine;

```sh
$ npm i
$ npm run norestart
// go to  http://localhost:3000/ & http://localhost:3000/api/movies
$ sudo kill -9 $(sudo lsof -t -i:3000) // will stop the server
```

2. For frontend install **LIVE SERVER in Visual Studio Code** & run it.
   Your &nbsp; /views/index.html will be displayed at http://localhost:5500/views/.
3. For the css & js links in your **index.html** use **absolute paths**, eg

   <link rel="stylesheet" href="/public/stylesheets/styles.css" >
   <script src="/public/javascripts/index.js"></script>

4. In **app.js**:
   // app.use(express.static(path.join(**dirname + '/views')));
   // app.use(express.static(path.join(**dirname, '/public')));

### Useful code snippets when using FETCH API

```sh
async function init() {
    try {
      const movies = await getMovies(fullUrl);
      renderMovies(movies);
    } catch (error) {
      console.error(
        'There has been a problem with your fetch operation: ' + error.message
      );
    }
}

async function getMovies(url = '') {
    const res = await fetch(url, {
      method: 'GET', // default
      mode: 'cors', // default
      headers: {
        'Content-Type': 'application/json'
      }
    });
    if (!res.ok) {
      return handleErrors(res);
    }

    const json = await res.json(); // parses JSON response into native JavaScript objects
    return json.movies;
  }


const res = await fetch(url, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json'
      },
      body: JSON.stringify(newMovie) // body data type must match "Content-Type" header
    });


 function handleErrors(res) {
    if (res.status === 404) {
      throw new Error('Ooops, page not found');
    }
    throw new Error(
      res.statusText +
        ' Network error (internet is down, or CORS is misconfigured on the server-side, or permission issues, etc.)'
    );
  }
```


changes
//app.set('views', path.join(__dirname, 'views'));
//app.use(express.static(__dirname + "/views"));

index route
res.sendFile('index.html');


app.js
var createError = require('http-errors');
var express = require('express');
var path = require('path');
var cookieParser = require('cookie-parser');
var logger = require('morgan');
var mongoose = require('mongoose');
var dbTest = require('./helpers/dbTest');

var indexRouter = require('./routes/index');
var movieRouter = require('./routes/movie');
var app = express();

// db
const dbName = 'mustwatch';
// uncomment to test mongo
// dbTest.test(mongoose)
mongoose.connect('mongodb://localhost:27017/' + dbName, {
  useNewUrlParser: true,
  useUnifiedTopology: true
});
mongoose.set('debug', true);
const conn = mongoose.connection;
conn
  .on('open', () => console.log('------- Connected to db ' + dbName))
  .catch(err => console.log(err));
mongoose.Promise = Promise;

app.use(logger('dev'));
app.use(express.json());
app.use(express.urlencoded({ extended: false }));
app.use(cookieParser());
// view engine setup
//app.use(express.static(path.join(__dirname + '/views')));
//app.use(express.static(path.join(__dirname, '/public')));

app.use('/', indexRouter);
app.use('/api/movies', movieRouter);

// catch 404 and forward to error handler
app.use(function(req, res, next) {
  next(createError(404));
});

// error handler
app.use(function(err, req, res, next) {
  // set locals, only providing error in development
  res.locals.message = err.message;
  res.locals.error = req.app.get('env') === 'development' ? err : {};

  // render the error page
  res.status(err.status || 500);
  res.send('error');
});

module.exports = app;


models/movie.js
var express = require('express');
var mongoose = require('mongoose');

// _id field by default 
const movieSchema = new mongoose.Schema({
  title: { type: String, default: 'No title', trim: true, required: true },
  genre: { type: String, default: 'Not specified', trim: true },
  released: { type: Number },
  isWatched: { type: Boolean, default: false },
  rating: { type: Number, default: 0 },
  createdAt: { type: Date, default: Date.now }
});
module.exports = mongoose.model('Movie', movieSchema);

/* 
   for postman or for db.movies.insertMany([.....])
    {"title": "Superman", "genre": "fantasy", "released": 2008, "isWatched": false, "rating": 0} 	
    {"title": "Star Wars", "genre": "sci-fi", "released": 1985, "isWatched": false, "rating": 0} 	
    {"title": "Thieves", "genre": "drama", "released": 2017, "isWatched": false, "rating": 0}  
 */


helpers/dbTest.js
#!/usr/bin/env node

exports.test = (mongoose) => {
  function log() {
    console.dir(arguments);
  }

  const opts = { useNewUrlParser: true, useUnifiedTopology: true };

  //const mongoose = require('mongoose');
  console.log(`VERSION: ${mongoose.version}`);
  const ver = mongoose.version.charAt(0);
  const conn = mongoose.connection;
  const Schema = mongoose.Schema;

  if (ver < 5) {
    opts.useMongoClient = true;
    mongoose.Promise = global.Promise;
  }
  // doesnt create the db immediately, only when you add sth to it
  mongoose.connect('mongodb://localhost/kurak', opts);

  const schema = new Schema({
    name: String
  });
  const Test = mongoose.model('test', schema);
  const test = new Test({ name: 'test' });

  async function run() {
    await conn.dropDatabase();
    await test.save();
    mongoose.set('debug', log);
    const count = await Test.find({ _id: test._id }).countDocuments();
    if (count === 1) {
      console.log('---- Mongo test passed ---- objects found: ' + count);
    } else {
      console.log('---- MONGO TEST FAILED ---- objects found: ' + count);
      console.log('---- !!! CLOSING the connection to mongo !!! --------');
      return conn.close();
    }
  }

  run();
};

module.exports = exports;


public/javascripts/index.js
window.addEventListener('DOMContentLoaded', event => {
  console.log('DOM fully loaded and parsed');

  const futureList = document.getElementById('futureList');
  const historyList = document.getElementById('historyList');
  const titleInput = document.getElementById('titleInput');

  const baseUrl = 'http://localhost:3000';
  const apiUrl = '/api/movies/';
  const fullUrl = baseUrl + apiUrl;

  /* ---------------------------------------------------------------- */
  init();

  async function init() {
    try {
      const movies = await getMovies(fullUrl);
      renderMovies(movies);
    } catch (error) {
      console.error(
        'There has been a problem with your fetch operation: ' + error.message
      );
    }
  }

  /* --------------- event listeners ---------------------------- */

  titleInput.addEventListener('keydown', async function fn(e) {
    var key = e.which || e.keyCode || 0;
    if (key == 13) {
      try {
        const movieTitle = titleInput.value;
        // add year & genre
        const newMovie = { title: movieTitle };
        titleInput.value = '';
        const savedMovie = await postMovie(fullUrl, newMovie);
        futureList.appendChild(createMovieItem(savedMovie));
      } catch (error) {
        console.error(
          'There has been a problem with your fetch operation: ' + error.message
        );
      }
    }
  });

  futureList.addEventListener('click', handleListClick);
  historyList.addEventListener('click', handleListClick);

  function handleListClick(e) {
    let clickedList = this === futureList ? futureList : historyList;
    let unclickedList = clickedList == historyList ? futureList : historyList;

    let listItem = getClosest(e.target, 'LI');
    const movieId = listItem.dataset.id;

    try {
      if (e.target.className === 'deleteBtn') {
        deleteMovie(movieId);
        clickedList.removeChild(listItem);
      } else {
        let isWatched = clickedList === futureList ? true : false;
        updateMovie(movieId, { isWatched });
        clickedList.removeChild(listItem);
        unclickedList.appendChild(listItem);
      }
    } catch (error) {
      console.error(
        'There has been a problem with your fetch operation: ' + error.message
      );
    }
  }

  /* ------------------------- API calls ---------------------- */

  async function getMovies(url = '') {
    const res = await fetch(url, {
      method: 'GET', // default
      mode: 'cors', // default
      headers: {
        'Content-Type': 'application/json'
      }
    });
    if (!res.ok) {
      return handleErrors(res);
    }

    const json = await res.json(); // parses JSON response into native JavaScript objects
    return json.movies;
  }

  async function postMovie(url = '', newMovie = {}) {
    const res = await fetch(url, {
      method: 'POST',
      mode: 'cors',
      headers: {
        'Content-Type': 'application/json'
      },
      body: JSON.stringify(newMovie) // body data type must match "Content-Type" header
    });

    if (!res.ok) {
      return handleErrors(res);
    }
    const movie = await res.json();
    return movie; // parses JSON response into native JavaScript objects
  }

  async function deleteMovie(movieId) {
    const res = await fetch(fullUrl + movieId, {
      method: 'DELETE'
    });
    if (!res.ok) {
      return handleErrors(res);
    }

    return await res.json();
  }

  async function updateMovie(movieId, { isWatched }) {
    const res = await fetch(fullUrl + movieId, {
      method: 'PUT',
      headers: {
        'Content-Type': 'application/json'
      },

      body: JSON.stringify({ isWatched })
    });

    if (!res.ok) {
      return handleErrors(res);
    }
    const movie = await res.json();
    return movie;
  }

  /* ----------------- view rendering ----------------------------*/

  function renderMovies(movies) {
    futureList.innerHTML = '';
    historyList.innerHTML = '';

    for (const movie of movies) {
      const movieElem = createMovieItem(movie);
      if (movie.isWatched) {
        historyList.appendChild(movieElem);
      } else {
        futureList.appendChild(movieElem);
      }
    }
  }

  function createMovieItem(movie) {
    const { _id, title, released, genre, rating, isWatched } = movie;
    let listItem = document.createElement('LI');
    listItem.setAttribute('data-id', _id);
    listItem.innerHTML = `<p class="movieItem">
        <strong>${title}</strong> 
       </p>`;
    /*  released in <span>${released}</span>,  <span>${genre}</span> */

    listItem.innerHTML += `<span class="deleteBtn">X</span>`;
    return listItem;
  }

  /* ----------------- helpers ----------------------------*/

  function getClosest(elem, selector) {
    for (; elem && elem !== document; elem = elem.parentNode) {
      if (elem.matches(selector)) return elem;
    }
    return null;
  }

  function handleErrors(res) {
    if (res.status === 404) {
      throw new Error('Ooops, page not found');
    }
    throw new Error(
      res.statusText +
        ' Network error (internet is down, or CORS is misconfigured on the server-side, or permission issues, etc.)'
    );
  }
});
